include/veriblock/blockchain/base_block_tree.hpp

```diff
@@ -520,6 +555,91 @@ struct BaseBlockTree {
   }
 
  protected:
+  //! Marks `block` as finalized.
+  //!
+  //! Final blocks can not be reorganized, thus we can remove outdated blocks
+  //! from tips_ and free some RAM by deallocating blocks past last final block
+  //! minus `N` (`preserveBlocksBehindFinal`) blocks. In ALT and VBK chains N
+  //! will equal to at least Endorsement Settlement Interval, so that
+  //! endorsements in blocks after finalized block will point to an existing
+  //! *endorsed* block.
+  //!
+  //! @param[in] block hash of block to be finalized. Must be part of active
+  //! chain.
+  //!
+  //! @warning This action is irreversible. You will need to reload the whole
+  //! tree to recover in case if this func is called by mistake.
+  //!
+  //! @returns false if block not found or prereq are not met
+  virtual bool finalizeBlockImpl(const hash_t& block,
+                                 // see config.preserveBlocksBehindFinal()
+                                 int32_t preserveBlocksBehindFinal) {
+    auto* index = getBlockIndex(block);
+    if (!index) {
+      return false;
+    }
+
+    // block is already final
+    if (index->finalized) {
+      return true;
+    }
+
+    // prereq is not met - finalized block must be on active chain
+    if (!activeChain_.contains(index)) {
+      return false;
+    }
+
+    // first, update active chain (it should start with
+    // 'index' but we also need to preserve `preserveBlocksBehindFinal` blocks
+    // before it). all outdated blocks behind `index` block will be deallocated
+    int32_t firstBlockHeight = index->getHeight() - preserveBlocksBehindFinal;
+    int32_t bootstrapBlockHeight = getRoot().getHeight();
+    firstBlockHeight = std::max(bootstrapBlockHeight, firstBlockHeight);
+    activeChain_ = Chain<index_t>(firstBlockHeight, activeChain_.tip());
+
+    // second, erase candidates from tips_ that will never be activated
+    erase_if<index_t>(tips_, [this, index](const index_t* tip) -> bool {
+      VBK_ASSERT(tip);
+
+      // active chain can not be outdated
+      if (activeChain_.contains(tip)) {
+        return false;
+      }
+
+      if (isBlockOutdated(*index, *tip)) {
+        return true;
+      }
+
+      return false;
+    });
+
+    // third, deallocate blocks that are outdated, if any
+    {
+      auto begin = blocks_.cbegin();
+      auto end = blocks_.cend();
+      for (auto it = begin; it != end;) {
+        auto* candidate = it->second.get();
+
+        bool outdated =
+            isBlockOutdated(*index, *candidate, preserveBlocksBehindFinal);
```

Comment:
Since isBlockOutdated uses getAncestor and getForkBlock, this has N*N complexuty